[Refactor/11] Refactoring do controller dos pets

damews opened the pull request:
Alteração do funcionamento da função salvar Pet. Agora, é construído a partir de um DTO. 
Recursos (getters/setters) foram necessários para poder incluir o pet na foto.
Repositories foram adicionados.

damews committed:
Alterando nome da função da busca geral

damews committed:
Merge branch 'release/2' into refactor/11

damews committed:
Novos repositories, alterando save e DTO

Versão inicial da nova manipulação dos dados para salvar no banco.

Foi adicionado DTO,  novos repositories.

damews committed:
Solução para salvar um pet

A solução (ainda inicial) ainda apresenta pontos de melhoria. A divisão entre classes que realizam este tipo de operações é necessária.

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -16,13 +21,19 @@
+    @Autowired
+    private FotoRepository fotoRepository;
+
     @GetMapping("/")
```
Dá pra tirar o `("/")` pra gente não ter que ficar adicionando o / no final da url pra fazer requisição

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -16,13 +21,19 @@
+
     @GetMapping("/")
-    private List<Pet> buscarTodos() {
+    private List<Pet> buscar() {
```
Listar todos os pets pode ser inviável à medida que o número de registros aumenta. O certo é paginar a busca. Dá uma olhada como a Helena fez no #17

LiserLine replied:
Feito.

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -16,13 +21,19 @@
     @GetMapping("/{id}")
-    public ResponseEntity<Pet> buscar(@PathVariable("id") UUID id) {
+    public ResponseEntity<Pet> buscarPorId(@PathVariable("id") UUID id) {
         return ResponseEntity.ok(petRepository.findById(id).get());
```
A gente precisa tratar o caso em que o ID não existe. Talvez criar uma exception customizada tipo "PetNaoCadastradoException" e retornar um código diferente pra ficar mais fácil pro front tratar.

LiserLine replied:
Retornando _notFound()_.
Discutir criações de exceções padronizadas.

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -32,8 +43,24 @@
     }
 
     @PostMapping("/")
```
Dá pra tirar o `("/") ` pra gente não ter que ficar colocando / no final da url pra requisição

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -32,8 +43,24 @@
+    public Pet salvar(@RequestBody PetDTO petDTO) {
+        Localizacao localizacao = localizacaoRepository.saveAndFlush(petDTO.getLocalizacao());
+
+        Pet pet = new Pet(petDTO.getId(), petDTO.getNome(),
```
Vê o negócio do `@Builder` do Lombok que eu comentei. Ficaria algo do tipo:
```java
Pet pet = Pet.builder()
    .id(petDTO.getId())
    .nome(petDTO.getNome())
...
   .build();
```

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -32,8 +43,24 @@
+                petDTO.getIdade(), petDTO.getUsuario(),
+                localizacao, petDTO.isAtivo());
+
+        petRepository.saveAndFlush(pet);
```
Você tá dando `saveAndFlush` mas não tá pegando o retorno. Assim, quanto você faz o `foto.setPet(pet)` ali embaixo você tá colocando na foto um pet que ainda não tem ID. Isso vai dar ruim na hora que ele for salvar, porque ele precisa do ID do pet pra cadastrar a foto.

LiserLine replied:
O pet no qual eu fiz a construção já não possui o ID que eu quero pra salvar as fotos? A questão de pegar o retorno do saveAndFlush é pra eu ter a garantia do ID ou algo do tipo? Ou porque o vai ser gerado um ID novo?

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -32,8 +43,24 @@
+        return pet;
     }
 
     @DeleteMapping("/{id}")
```
A gente decidiu que o Pet e o Usuário não seriam deletados do banco, só inativados. Pode mudar aqui, por favor? Usa esse ID pra buscar o pet, altera o status e atualiza ele no banco. :smile:

rartner commented on src/main/java/br/com/academiadev/thunderpets/dto/PetDTO.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.UUID;
+
+@Getter
+@Setter
```
Você pode substituir essas duas anotações e deixar só o `@Data`. Segundo a documentação do Lombok, o `@Data` já contem getters, setters, equals e hashcode.

rartner commented on src/main/java/br/com/academiadev/thunderpets/dto/PetDTO.java:
```diff
@@ -0,0 +1,66 @@
+public class PetDTO {
+
+    @Id
+    @GeneratedValue
```
Como é um DTO, não precisa dessas anotações. Elas só são necessárias nas entidades pra fazer a validação no banco de dados. Aqui você pode só declarar os campos sem nenhuma anotação.
Talvez precise das anotações nos enums, mas não tenho certeza. O resto não precisa.

LiserLine replied:
Acabei por remover os enums também. Creio que como essas anotações são necessárias apenas nos modelos (posteriormente na validação no banco de dados) e o DTO é apenas um cara para lidar com todas as info que vem do front, não seja necessário colocá-las.

rartner commented on src/main/java/br/com/academiadev/thunderpets/model/Foto.java:
```diff
@@ -10,6 +9,8 @@
 @Entity
 @NoArgsConstructor
+@Getter
+@Setter
```
Como aqui já tem o `@Data`, não precisa do `@Getter` e `@Setter`

rartner commented on src/main/java/br/com/academiadev/thunderpets/repository/FotoRepository.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.UUID;
+
+public interface FotoRepository extends JpaRepository<Foto, UUID> {
+}
```
Poderia deixar uma linhazinha em branco nos repositórios sem declaração de método, só pra manter o padrão? :heart:

damews committed:
Retirando coisas desnecessárias | Formatação de código.

Substituindo anotações;
Formatação de código;
Arrumando classe PetDTO.

damews committed:
Alterando funcionalidade [delete] do PetController

Não é necessário excluir o cadastro do banco, apenas o deixar inativo.

damews committed:
Utilização do Builder Patter no salvar | Retorno do pet

damews committed:
Mapeamento de pets (DTO) | Busca por ID melhorada

Feito mapeamento da classe Pet para PetDTO.
Busca por ID agora lida com IDs não existentes, dando o devido retorno.

damews committed:
Paginação de pets | Tratamento de exceção no buscarPorId

Ao invés de retornar uma lista imensa dos pets cadastrados, agora, retorna a uma página específica de pets.
---
Tratamento de exceçõo para pets não cadastrados [Ver criação de exceções padronizadas]

damews committed:
Checkstyle [1]

damews committed:
Mapper | Exception | Refatorando funções

Adicionado um mapper para o PetDto;
Adicionado um Exception para pet não encontrado;
Refatorando funções para utilizar as novas classes.

damews committed:
Conversão no petMapper | Aplicando a conversão adicionada

damews committed:
Circle CI bugado

rartner commented on src/main/java/br/com/academiadev/thunderpets/controller/PetController.java:
```diff
@@ -16,32 +29,68 @@
+                             @RequestParam(defaultValue = "ASC") Sort.Direction direcao,
+                             @RequestParam(defaultValue = "nome") String campoOrdenacao) {
+        PageRequest paginacao = PageRequest.of(paginaAtual, tamanho, direcao, campoOrdenacao);
+        Page<Pet> paginaPets = petRepository.findAll(paginacao);
```
Precisamos adicionar um filtro pra retornar só os ativos

damews committed:
Circle CI - Arrumando

damews committed:
Anotações | Paginação Pets

Novas anotações foram adicionadas para a busca de pets. Agora é possível escolher entre pets ativos ou não. O padrão de busca agora é pela data de registro.

Novas anotações foram adicionadas para melhor funcionamento.

damews committed:
Circle CI - Arrumando 2

damews committed:
Anotação @AllArgsConstructor em PetDTO

damews committed:
Circle CI - Arrumando 3

damews committed:
Circle CI - Arrumando 4

Não aguento mais.
